src/backend/distributed/utils/statistics_collection.c

```diff
@@ -114,8 +124,160 @@ CollectBasicUsageStatistics(void)
+void
+CheckForUpdates(void)
+{
+	StringInfo url = makeStringInfo();
+	appendStringInfoString(url, STATS_COLLECTION_HOST "/v1/releases/latest?edition=");
+
+	if (!UrlEncode(url, CITUS_EDITION))
+	{
+		ereport(WARNING, (errmsg("url encoding '%s' failed", CITUS_EDITION)));
+	}
+
+	if (!SendHttpGetRequest(url->data, HTTP_TIMEOUT_SECONDS, &CheckForUpdatesCallback))
+	{
+		ereport(WARNING, (errmsg("checking for updates failed")));
+	}
+}
+
+
+/*
+ * CheckForUpdatesCallback receives the response for the request sent by
+ * CheckForUpdates(). It processes the response, and if there is a newer release
+ * of Citus available, logs a LOG message. This function returns 0 if there are
+ * any errors in the received response, which means we didn't consume the data.
+ * Otherwise, it returns (size * count) which means we consumed all of the data.
+ */
+static size_t
+CheckForUpdatesCallback(char *contents, size_t size, size_t count, void *userData)
+{
+	const int32 citusVersionMajor = CITUS_VERSION_NUM / 10000;
+	const int32 citusVersionMinor = (CITUS_VERSION_NUM / 100) % 100;
+	const int32 citusVersionPatch = CITUS_VERSION_NUM % 100;
+	Jsonb *responseJsonb = 0;
+	StringInfo releaseVersion = makeStringInfo();
+	int32 releaseMajor = 0;
+	int32 releaseMinor = 0;
+	int32 releasePatch = 0;
+	char *updateType = NULL;
+
+	StringInfo responseNullTerminated = makeStringInfo();
+	appendBinaryStringInfo(responseNullTerminated, contents, size * count);
+
+	/* jsonb_in can throw errors */
+	PG_TRY();
+	{
+		responseJsonb = DatumGetJsonb(
+			DirectFunctionCall1(jsonb_in, CStringGetDatum(responseNullTerminated->data)));
+	}
+	PG_CATCH();
+	{
+		FlushErrorState();
+		return 0;
+	}
+	PG_END_TRY();
+
+	if (!JsonbFieldStr(responseJsonb, "version", releaseVersion) ||
+		!JsonbFieldInt(responseJsonb, "major", &releaseMajor) ||
+		!JsonbFieldInt(responseJsonb, "minor", &releaseMinor) ||
+		!JsonbFieldInt(responseJsonb, "patch", &releasePatch))
+	{
+		return 0;
+	}
+
+	if ((releaseMajor > citusVersionMajor) ||
+		(releaseMajor == citusVersionMajor && releaseMinor > citusVersionMinor))
+	{
+		updateType = "major";
+	}
+	else if (releaseMajor == citusVersionMajor &&
+			 releaseMinor == citusVersionMinor &&
+			 releasePatch > citusVersionPatch)
+	{
+		updateType = "patch";
+	}
+
+	if (updateType != NULL)
+	{
+		ereport(LOG, (errmsg("a newer %s release of Citus (%s) is available",
+							 updateType, releaseVersion->data)));
+	}
+
+	return size * count;
+}
+
+
+/*
+ * JsonbFieldInt sets the given output variable to the integer value of the given
+ * field in the given JSONB object. If the field doesn't exist or its value is
+ * not an integer, this function returns false.
+ */
+static bool
+JsonbFieldInt(Jsonb *jsonb, const char *fieldName, int32 *result)
+{
+	JsonbValue *fieldValue = NULL;
+	JsonbValue key;
+	memset(&key, 0, sizeof(key));
+	key.type = jbvString;
+	key.val.string.len = strlen(fieldName);
+	key.val.string.val = pstrdup(fieldName);
```

Comment:
I could probably cast `fieldName` to `char *` instead of `pstrdup()`, but that didn't look very good.

Reply:
Done.